### ISSUE TITLE
Practice and staff member survey date import vs. manual entry

### ISSUE THREAD

**Issue by squabwrangler** [seemack/healthy-hearts-northwest-dev#57]
survey response information is currently tracked in an external spreadsheet which makes the information inaccessible to this application. Implement an import of the tracking excel spreadsheet data so that the information is available for display.

**Comment by rpardee** [comment]
Additional context: this request was originally made by Elena (owner of survey solicitation process) & put off by Roy as difficult to implement and possibly better managed in the app.

Now we have an additional request from Managers (Ross) to have practice & staff survey response information available for display to coaches.

**Comment by ltuz** [comment]
Here's Ross' request (not sure why he didn't say anything when he was on the phone with Erika and Chris today!)....

I have been building an excel display for coaches on the status of their various surveys. But it would be useful to have something like this built into the database so coaches can see all of their clinics' data in one place. The data they are looking for with surveys are returned yes/no, number of reminders, %returned, etc. I recognize the challenge of doing that on your end, though, given the challenges I've had building it for us.

**Comment by seemack** [comment]
It would be great to see a copy of this spreadsheet.  And separately to see
if ORPRN is doing something similar, so if we build something like this,
it'll hit everything both sites want to track.

Maybe next week or soon we could see how this (and if there are similar
reporting requests) fit into the overall project needs for the next 6
months or so.

**Comment by rpardee** [comment]
I just saved the sample xlsx that Elena sent us to \\groups\data\ctrhs\inst\AHRQ Healthy Hearts Northwest\DATA\Database\sample_survey_tracker.xlsx

**Comment by rpardee** [comment]
Ross sent via e-mail (sent to me by Leah) 23-feb-2016 a spreadsheet that *he* has been managing and sending to his coaches, which looks like this:

![image](https://cloud.githubusercontent.com/assets/26327/13333077/02c97582-dbba-11e5-9183-5bab2b784f0f.png)

That lovely formatting is very doable in the app, once we have the desired information in our db.

I'm assuming 'primary facilitator'  = assigned coach, so that & practice name are no problem--they're already in the db.

I believe these guys are in Elena's spreadsheet:

1. practice survey issued
2. practice survey returned

I'm not sure where this info would be--some other spreadsheet maybe?

1. staff survey issued
2. staff survey return rate
3. date welcome visit scheduled (app has welcome visits that have been recorded, but I don't think we have future dates for scheduled visits).
4. ready for ABCS survey (maybe this is derivable from info that's already in the db?)

So we'll need to 

1. add fields (and maybe some tables) in the underlying db to hold all 6 bits of information.
2. create a means by which that information can be updated by users--some combination of XLSX upload + in-app web form
3. create the display

One issue with importing Elena's and/or Ross' spreadsheets are that neither has a coded practice identifier.  It's not difficult to match by practice name, but only exact matches will match--so for instance right now the app has a practice named "Lopez Island Medical Clinic" (that's database id 18108, not sure what the studyid/practiceid is).  Ross' spreadsheet has one called "Island - Lopez Island Medical Clinic".  Humans can (probably) tell that this is the same clinic (I think?).  But the app won't know that--it just won't find a match, and won't be able to use the incoming spreadsheet row.  So optimally any to-be-imported spreadsheet will add a column to store either of those coded identifiers so there's less chance of user-introduced error.

**Comment by kuoxes1** [comment]
From practice survey tracker:
Practice ID
Date PS sent
Date PS returned
Notes (missing data)

from staff/ABCS survey tracker
Practice OD
Date staff survey sent
Staff survey return rate
OK for ABCS 
Q4 2015 sent
Q4 2015 returned
Q1 2016 sent
Q1 2016 returned
etc.

we don't track welcome visit scheduled, and welcome letter issued

We prefer a display

Practice survey is 3x, Staff member survey is 3x, abcs survey is 10x

**Comment by rpardee** [comment]
Elena--I'm not seeing a date returned field in the sample spreadsheet you sent a while ago (which I stashed at \\groups\data\ctrhs\inst\AHRQ Healthy Hearts Northwest\DATA\Database\sample_survey_tracker.xlsx).  Are you misremembering what was there, or are you changing your spreadsheet, or ...?  If you are changing, can you send a new sample xlsx?  I will also need a sample staff/ABCS xlsx to code against--can you put one in that /data/database dir as well please?

**Comment by seemack** [comment]
I'm tempted to split this issue into two, one for getting Staff Survey data into the database, and another that's specific to the requested report.  The solutions to one don't necessarily affect the solutions to the other.  Are folks alright with this, or is this level of nitpick just proof that I'm losing my mind?

**Comment by rpardee** [comment]
My inclination is to keep this one issue at least until I get an answer to my question above.  I am stalled until I get that (and even more importantly, sample spreadsheets).

**Comment by ltuz** [comment]
I agree with Roy to keep this issue as one. Elena needs to be the one to respond to Roy's question.

**Comment by kuoxes1** [comment]
Hi, I had misremembered, 
For the practice survey we are not tracking date returned, just completed or not

For the staff member survey, we  have a column for "% complete"

**Comment by rpardee** [comment]
Cool--thanks Elena.  As soon as I have sample spreadsheets I can start working again.

**Comment by seemack** [comment]
Just tossing out a possible change in strategy.  SurveyMonkey has an API that could allow us to programmatically pull data, and we could likely integrated it into our system that way. This may only work if the survey isn't changed - when I last was reading SurveyMonkey data (a couple years ago now) each change made such dramatic changes to the data model that I had to rewrite my code every time.

[Survey Monkey API reference](https://developer.surveymonkey.com/)

**Comment by rpardee** [comment]
That could definitely be slick--I'd love to see it.  Looks like there's a rubygem for interacting w/sm (https://github.com/hakamadare/rubygem-surveymonkey).

**Comment by kuoxes1** [comment]
elena will send spreadsheets to Roy ad Chris to review and also the log in for surveymonkey.com
The priority is for practice survey, we'd love for date sent, complete yes/no and maybe missing data field (not currently in there) to be imported

**Comment by rpardee** [comment]
For the record--Elena made good on her promise & coughed up a sample spreadsheet.

From the ensuing e-mail discussion:

> It’d be lovely if instead of using the heading “Staffsurvey_sent” you just used “survey_sent” so that field can be generic across the Practice & ABCS surveys as well (unless you intend to use “Staffsurvey_sent” in the Practice & ABCS spreadsheets as well, in which case nevermind).
> 
> So far I’ve got it set up to show you something like this upon a successful import:
> 
> File uploaded:	ORPRN - Staff Member Surveys Contact Sheet.xlsx
> Survey:	Staff, 1st administration
> New records created:	40
> Existing records updated:	1
> Rows unable to process:	0
> Error Messages	
> File Info:	File: RackMultipart20160308-11468-bgl9wz.xlsx
> Number of sheets: 1
> Sheets: Sheet1
> Sheet 1:
>   First row: 1
>   Last row: 42
>   First column: A
>   Last column: AL
> 
> You’ll want to pay close attention to those new & updated record counts & make sure you get the total N you expect.  I was surprised to see 1 ‘existing record updated’ on importing your sample data into a db w/no survey recs at all until I figured out that you’ve got a typo—the practice id 21124 is repeated.  The first time that id is encountered the import routine creates the record, and the second time it finds the existing record & overwrites its data w/what’s on the second.
> 
> From: Pardee, Roy 
> Sent: Tuesday, March 08, 2016 11:19 AM
> To: Kuo, Elena; Mack, Christopher
> Cc: Tuzzio, Leah
> Subject: RE: Practice Survey example dataset
> 
> Wunderbar—many thanks.  I’ll get on this & let you know what roadblocks I run into.
> 
> From: Kuo, Elena 
> Sent: Wednesday, March 02, 2016 10:47 PM
> To: Pardee, Roy; Mack, Christopher
> Cc: Tuzzio, Leah
> Subject: Practice Survey example dataset
> 
> Hi Roy,
> 
> Attached is a sample Staff Member Survey excel tracker.
> Three columns are highlighted in yellow that are needed for the import.
> Practice ID (to match)
> Date sent (MM/DD/YYYY currently)
> % returned (in % format currently)
> 
> For the Practice Survey it will be the same format, except it will be
> Practice ID (to match)
> Date sent (MM/DD/YYYY currently)
> Returned (complete if returned, not a date)
> 
> If you want columns to be named a certain way let us know and we can change it.
> or in the returned field for Practice Survey have it say "1" or something else instead of "complete", just let us know what format you prefer 
> 
> Elena
>

**Comment by rpardee** [comment]
More off-github conversation below.  Just FYI--I don't know how to continue work on this w/out understanding your spreadsheet Elena.


From: Pardee, Roy 
Sent: Wednesday, March 09, 2016 1:38 PM
To: Kuo, Elena; Mack, Christopher
Cc: Tuzzio, Leah
Subject: RE: Practice Survey example dataset

So some of these rows will signify ABCS and others signify Staff?  I’m not seeing any references to ABCS…

From: Kuo, Elena 
Sent: Tuesday, March 08, 2016 9:26 PM
To: Pardee, Roy; Mack, Christopher
Cc: Tuzzio, Leah
Subject: RE: Practice Survey example dataset

Here's one - the "real" one for upload would just have an extra column with Practice ID, which I still have yet to get to!  One of these days...
________________________________________
From: Pardee, Roy
Sent: Tuesday, March 08, 2016 6:04 PM
To: Kuo, Elena; Mack, Christopher
Cc: Tuzzio, Leah
Subject: RE: Practice Survey example dataset
Okay, cool—can you give me a test file w/those combined?  Doesn’t have to be real data…
 
From: Kuo, Elena 
Sent: Tuesday, March 08, 2016 5:30 PM
To: Pardee, Roy; Mack, Christopher
Cc: Tuzzio, Leah
Subject: RE: Practice Survey example dataset
 
Ok let me know if any other headers need to be changed besides survey_sent in both.  ABCS and Staff member surveys are tracked in the same spreadsheet so we'll need something to delineate between them :)

**Comment by rpardee** [comment]
Ah, that is a different spreadsheet.  So you mean to have one row per practice, with both staff and ABCS signified on the same row, but in different columns.

So the column headed "Sent" would hold the date the staff survey was sent, and the column headed "% of Staff completed" would of course hold the percent of staff that have completed the staff survey.

Which column holds the date the ABCS survey was sent?
How can we tell which practices have responded to (returned?) the ABCS?

When it's time for the second staff survey, what will your spreadsheet look like?  Will you start new spreadsheets for those?  How about successive ABCSs?

-------------------------------------------
From: Kuo, Elena 
Sent: Saturday, March 12, 2016 1:13 PM
To: Pardee, Roy
Cc: Mack, Christopher
Subject: RE: [healthy-hearts-northwest-dev] Practice and staff member survey date import vs. manual entry (#57)

Here is a spreadsheet again, perhaps I sent the wrong version.  ABCS stuff is at the end of the spreadsheet, around column AC.  Please let me know if you need column headers renamed!  There is 1 spreadsheet for QUALIS and one for ORPRN staff/ABCS survey tracking.

**Comment by kuoxes1** [comment]
**Spreadsheet 1:**
Practice survey : 1 time only into database variables: sent, returned

**Spreadsheet2:**
Staff member survey:  1 time only into database variable: 75% complete
ACBS survey:  quarterly upload  variable: completed

**Comment by rpardee** [comment]
Made some good progress on this--working out w/Chris how/when to push it out to prod for evaluation.

**Comment by rpardee** [comment]
Here's a preview of what will be a /surveys view:

![image](https://cloud.githubusercontent.com/assets/26327/13894777/fe8a0240-ed28-11e5-8b8e-ebd8e7a8c02e.png)

And here's an option for the coach practice view w/info on who has completed the Practice survey:
![image](https://cloud.githubusercontent.com/assets/26327/13894768/eb213138-ed28-11e5-88c0-81b0c26d1989.png)

**Comment by rpardee** [comment]
Today's db meeting consisted of Elena & me looking at the spreadsheet upload feature.  It looked to work well, with the one exception of the "% of Staff completed" column in the staff/abcs spreadsheet type not successfully uploading.  I'm looking into that now.

**Comment by rpardee** [comment]
The issue was that there was a trailing space in the column with the percent complete values in the file we tried uploading this morning.  The import routine tried to pull the value out of a column headed:

`"% of Staff completed"`

but what was there was:

`"% of Staff completed "`

I elaborated that part of the import to look for either header label.

**Comment by seemack** [comment]
I'm opening this back up so Roy can take a look - I've modified his code and want his input before calling this done.

The coach list was bombing in dev on records that had a practice survey.  This was the cause:
```
<% if practice.practice_surveys.exists? %>
  <% if practice.practice_survey.status == "complete" %>
    Yes
  <% else %> 
    No--practice survey not yet sent.
  <% end %>
<% else %>
  No--practice survey sent, but not yet returned.
<% end %>
```

Practice_surveys was old work that got 80% of the way done with building the Practice Survey before Elena built it on Survey Monkey.  I'm pretty sure we want just Survey instead.  Then because Practice:PracticeSurvey is a 1:many relationship, there's no practice.practice_survey method.  I've changed that to practice.practice_surveys.first, **assuming that there will only be one survey (or that the first one is the most important).**  If that's not the case, let's revisit this.
